homework_lection2_sergiy_baranov/main.py

```diff
@@ -0,0 +1,128 @@
+import json
+from pydantic import BaseModel
+from abc import ABC
+
+# Клас "Книга"
+class BookModel(BaseModel):
+    title: str
+    author: str
+    year_published: int
+
+    def __str__(self):
+        return f"Title: {self.title}, Author: {self.author}, Year Published: {self.year_published}"
+
+    def to_dict(self):
+        return {
+            "title": self.title,
+            "author": self.author,
+            "year_published": self.year_published
+        }
+
+
+class AbsLib(ABC):
+    def add_book(self, book):
+        pass
+    def remove_book(self, book):
+        pass
+    def books_by_author(self, book):
+        pass
+
+# Клас "Бібліотека"
+class Library(AbsLib):
+    def __init__(self):
+        self.books = []
+
+    def add_book(self, book):
+        self.books.append(book)
+
+    def remove_book(self, book):
+        if book in self.books:
+            self.books.remove(book)
+
+    def __iter__(self):
+        self.index = 0
+        return self
+
+    def __next__(self):
+        if self.index < len(self.books):
+            book = self.books[self.index]
+            self.index += 1
+            return book
+        else:
+            raise StopIteration
+
+    def books_by_author(self, author):
+        return [book for book in self.books if book.author == author]
+
+# Декоратори
+def log_add_book(func):
+    def wrapper(self, book):
+        print(f"Adding book: {book.title} by {book.author}")
+        return func(self, book)
+    return wrapper
+
+def check_book_existence(func):
+    def wrapper(self, book):
+        if book in self.books:
+            return func(self, book)
+        else:
+            print(f"Book not found in the library: {book.title} by {book.author}")
+    return wrapper
+
+# Контекстний менеджер
+class LibraryFileManager:
```

Comment:
I see such context manager in second homework. It works, but does not what it should. On enter it reads file and on exit closes. I expected to see similar like python's built-in CM open() works. You can do the same.